geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/AbstractFileSystemStorage.scala

```diff
@@ -187,32 +211,87 @@ abstract class AbstractFileSystemStorage(
           logger.error(s"Failed to delete some files: [${failures.mkString(", ")}]")
         }
 
-        logger.debug(s"Compacted $written records into file $path")
+        logger.debug(s"Compacted $written records")
       }
     }
   }
 
   /**
-    * Create a new writer
-    *
-    * @param partition partition being written to
-    * @param action write type
-    * @return
-    */
-  private def createWriter(partition: String, action: StorageFileAction): FileSystemWriter = {
-    val fileType = action match {
-      case StorageFileAction.Append => FileType.Written
-      case StorageFileAction.Modify => FileType.Modified
-      case StorageFileAction.Delete => FileType.Deleted
-      case _ => throw new NotImplementedError(s"Unexpected storage action type: $action")
+   * Create a new writer
+   *
+   * @param partition partition being written to
+   * @param action write type
+   * @param fileType file type
+   * @param targetFileSize target file size
+   * @return
+   */
+  private def createWriter(
+      partition: String,
+      action: StorageFileAction,
+      fileType: FileType,
+      targetFileSize: Option[Long] = None): FileSystemWriter = {
+
+    def pathAndObserver: WriterConfig = {
+      val path = StorageUtils.nextFile(context.root, partition, metadata.leafStorage, extension, fileType)
+      PathCache.register(context.fc, path)
+      val updateObserver = new UpdateObserver(partition, path, action)
+      val observer = if (observers.isEmpty) { updateObserver } else {
+        new CompositeObserver(observers.map(_.apply(path)).+:(updateObserver))
+      }
+      WriterConfig(path, observer)
+    }
+
+    targetSize(targetFileSize) match {
+      case None => createWriter(pathAndObserver)
+      case Some(s) => new ChunkedFileSystemWriter(Iterator.continually(pathAndObserver), estimator(s))
+    }
+  }
+
+  private def createWriter(config: WriterConfig): FileSystemWriter = createWriter(config.path, config.observer)
+
+  /**
+   * Writes files up to a given size, then starts a new file
+   *
+   * @param paths iterator of files to write
+   * @param estimator target file size estimator
+   */
+  class ChunkedFileSystemWriter(paths: Iterator[WriterConfig], estimator: FileSizeEstimator)
+      extends FileSystemWriter {
+
+    private var count = 0L // number of features written
+    private var total = 0L // sum size of all finished chunks
+    private var remaining = estimator.estimate(0L)
+
+    private var path: Path = _
+    private var writer: FileSystemWriter = _
+
+    override def write(feature: SimpleFeature): Unit = {
+      if (writer == null) {
+        val config = paths.next
+        path = config.path
+        writer = createWriter(config)
+      }
+      writer.write(feature)
+      count += 1
+      remaining -= 1
+      if (remaining == 0) {
```

Comment:
Just to make sure I'm reading this correctly... we are only updating the estimate after we've written how many ever records that the estimator initially suggested that we write, correct?

Reply:
The question I have is if we should check earlier or more often?

Reply:
in the case of parquet and orc, we can't actually check the file size until the file is closed, as there is a bunch of metadata/overhead that gets written at that point. but to answer your broader question, there's a trade-off between the getting the file the exact size and the cost of checking the file size as you go, which requires some I/O. so the approach taken here is to estimate an initial number of features, and then check the file size after that, and refine the estimate for the next iteration. the fs storage system will also persist the estimate (unless you disable it) so that files will stay a consistent size between invocations.